subsys/pm/device_runtime.c

```diff
@@ -127,71 +128,93 @@ static int pm_device_request(const struct device *dev,
 						  PM_DEVICE_STATE_SUSPEND,
 						  NULL, NULL);
 		}
-		ret = 0;
 		goto out;
 	}
 
-	(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+	(void)k_mutex_lock(&dev->pm->lock, K_FOREVER);
+
+	if (!dev->pm->enable) {
+		ret = -ENOTSUP;
+		goto out_unlock;
+	}
+
+	if (target_state == PM_DEVICE_STATE_ACTIVE) {
+		dev->pm->usage++;
+		if (dev->pm->usage < 0) {
+			goto out_unlock;
+		}
+	} else {
+		dev->pm->usage--;
+		if (dev->pm->usage > 1) {
+			goto out_unlock;
+		}
+	}
+
 
 	/* Return in case of Async request */
 	if (pm_flags & PM_DEVICE_ASYNC) {
-		ret = 0;
-		goto out;
+		(void)k_work_schedule(&dev->pm->work, K_NO_WAIT);
+		goto out_unlock;
 	}
 
-	k_mutex_init(&request_mutex);
-	k_mutex_lock(&request_mutex, K_FOREVER);
-	(void)k_condvar_wait(&dev->pm->condvar, &request_mutex, K_FOREVER);
-	k_mutex_unlock(&request_mutex);
+	pm_device_runtime_state_set(dev->pm);
```

Comment:
everything here is going to be run synchronously, so we do not need to lock/unlock the mutex in both `pm_device_runtime_state_set` and the callback which will be covered by the existing mutex that was taken above.

Reply:
pm_device_runtime_state_set is also used by the workqueue, that is why there is a mutex there, as mutex are re-entrant, it is not a problem lock() twice from the same thread (sync path).  Alos,v the callback was not assuming that would run synchronous. Again, in the end all requested changes are based on the assumption that devices do it synchronously.

Reply:
agree about mutexes holding a count, but it is still another call that needs to be made which we could save. One more important point to note when we get to async, we can't use mutexes in async calls, mutexes can't be unlocked in ISR context.